packages/astro/src/default/utils/content.ts

```diff
@@ -75,250 +208,226 @@ export async function getTutorial(): Promise<Tutorial> {
+          throw new Error(`Could not find chapter '${chapterId}'`);
+        }
+
+        lesson.chapter = {
+          id: chapterId,
+          title: tutorial.parts[partId].chapters[chapterId].data.title,
+        };
+      }
+
+      tutorial.lessons.push(lesson);
     }
   }
 
   if (!tutorialMetaData) {
     throw new Error(`Could not find tutorial 'meta.md' file`);
   }
 
-  // let's now compute the order for everything
-  const partsOrder = getOrder(tutorialMetaData.parts, _tutorial.parts);
-
-  for (let p = 0; p < partsOrder.length; ++p) {
-    const partId = partsOrder[p];
-    const part = _tutorial.parts[partId];
+  return { tutorial, tutorialMetaData };
+}
 
-    if (!part) {
-      logger.warn(`Could not find '${partId}', it won't be part of the tutorial.`);
-      continue;
-    }
+function getOrder(
+  order: string[] | undefined,
+  fallbackSourceForOrder: Record<string, Part | Chapter> | Lesson['id'][],
+): string[] {
+  if (order) {
+    return order;
+  }
 
-    if (!_tutorial.firstPartId) {
-      _tutorial.firstPartId = partId;
-    }
+  const keys = Array.isArray(fallbackSourceForOrder)
+    ? [...fallbackSourceForOrder]
+    : Object.keys(fallbackSourceForOrder);
 
-    part.order = p;
+  // default to an order based on having each folder prefixed by their order: `1-foo`, `2-bar`, etc.
+  return keys.sort((a, b) => {
+    const numA = parseInt(a, 10);
+    const numB = parseInt(b, 10);
 
-    const chapterOrder = getOrder(part.data.chapters, part.chapters);
+    return numA - numB;
+  });
+}
 
-    for (let c = 0; c < chapterOrder.length; ++c) {
-      const chapterId = chapterOrder[c];
-      const chapter = part.chapters[chapterId];
+function sortCollection(collection: CollectionEntryTutorial[]) {
+  return collection.sort((a, b) => {
+    const depthA = a.id.split('/').length;
+    const depthB = b.id.split('/').length;
 
-      if (!chapter) {
-        logger.warn(`Could not find '${chapterId}', it won't be part of the part '${partId}'.`);
-        continue;
-      }
+    return depthA - depthB;
+  });
+}
 
-      if (!part.firstChapterId) {
-        part.firstChapterId = chapterId;
-      }
+function getSlug(entry: CollectionEntryTutorial) {
+  let slug: string = entry.slug;
 
-      chapter.order = c;
+  if (entry.slug.includes('/')) {
+    const parts = entry.slug.split('/');
+    const _slug = parts.at(-2);
 
-      const lessonOrder = getOrder(chapter.data.lessons, chapter.lessons);
+    if (!_slug) {
+      throw new Error('Invalid slug');
+    }
 
-      for (let l = 0; l < lessonOrder.length; ++l) {
-        const lessonId = lessonOrder[l];
-        const lesson = chapter.lessons[lessonId];
+    slug = _slug;
+  }
 
-        if (!lesson) {
-          logger.warn(`Could not find '${lessonId}', it won't be part of the chapter '${chapterId}'.`);
-          continue;
-        }
+  return slug;
+}
 
-        if (!chapter.firstLessonId) {
-          chapter.firstLessonId = lessonId;
-        }
+function resolveIds(
```

Comment:
Nice, this function is pretty cool, makes the logic focused in a single function.

Reply:
And this is intentionally verbose 😄